mlkem/indcpa.c

```diff
@@ -166,41 +168,86 @@ static unsigned int rej_uniform(int16_t *r,
 // Not static for benchmarking
 void gen_matrix(polyvec *a, const uint8_t seed[KYBER_SYMBYTES], int transposed)
 {
-    unsigned int ctr, i, j, k;
-    unsigned int buflen, off;
-    uint8_t buf[GEN_MATRIX_NBLOCKS * XOF_BLOCKBYTES + 2];
+    unsigned int ctr, i, j;
+    unsigned int buflen;
+    uint8_t bufx[KECCAK_WAY][GEN_MATRIX_NBLOCKS * XOF_BLOCKBYTES];
+    uint8_t *buf = NULL;
+    int16_t *vec = NULL;
+    uint8_t x, y;
     xof_state state;
 
-    for (i = 0; i < KYBER_K; i++)
+    keccakx4_state statex;
+    // The input data to each Keccak lane.
+    // Original size; KYBER_SYMBYTES + 2, we add padding to make align load/store.
+    uint8_t seedxy[KECCAK_WAY][KYBER_SYMBYTES + 16];
+
+    for (i = 0; i < (KYBER_K * KYBER_K / KECCAK_WAY) * KECCAK_WAY; i += KECCAK_WAY)
     {
-        for (j = 0; j < KYBER_K; j++)
+        for (j = 0; j < KECCAK_WAY; j++)
         {
+            memcpy(seedxy[j], seed, KYBER_SYMBYTES);
```

Comment:
This can be hoisted out of the loop?

Reply:
You're right, I overlook this line. I moved it out of the loop.